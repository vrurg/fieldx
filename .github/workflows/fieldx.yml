name: CI

on:
  push:
  pull_request:

env:
  CARGO_TERM_COLOR: always
  CARGO_INCREMENTAL: 0

jobs:
  with_features:
    runs-on: ${{ matrix.os }}-latest
    strategy:
      matrix:
        os: [ubuntu, macos, windows]
        toolchain:
          - 1.77
          - stable
          - beta
          - nightly
        feature: [serde]
    steps:
      - uses: actions/checkout@v4
      - run: rustup update ${{ matrix.toolchain }} && rustup default ${{ matrix.toolchain }}
      - run: cargo build --verbose --all --features ${{ matrix.feature }}
      - run: cargo test --verbose --all --features ${{ matrix.feature }}
      - run: cargo test --verbose --all --features diagnostics
        if: matrix.toolchain == 'nightly'
      - run: cargo test --verbose --all --all-features
        if: matrix.toolchain == 'nightly'
  no_features:
    runs-on: ${{ matrix.os }}-latest
    strategy:
      matrix:
        os: [ubuntu, macos, windows]
        toolchain:
          - 1.77
          - stable
          - beta
          - nightly
    steps:
      - uses: actions/checkout@v4
      - run: rustup update ${{ matrix.toolchain }} && rustup default ${{ matrix.toolchain }}
      - run: cargo build --verbose --all
      - run: cargo test --verbose --all
  build_docs:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - run: rustup default nightly
      - run: cargo install cargo-docs-rs
      - run: cargo docs-rs -p fieldx

