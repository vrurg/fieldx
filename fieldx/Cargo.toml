[package]
    authors.workspace      = true
    categories.workspace   = true
    description            = "Procedural macro for constructing structs with lazily initialized fields, builder pattern, and serde support with a focus on declarative syntax."
    edition.workspace      = true
    exclude.workspace      = true
    keywords.workspace     = true
    license-file.workspace = true
    name                   = "fieldx"
    readme                 = "../README.md"
    repository.workspace   = true
    version                = "0.1.8-beta.2"

    # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
    fieldx_aux    = { workspace = true }
    fieldx_derive = { workspace = true, features = ["tracing"] }
    getset        = { workspace = true }
    parking_lot   = { workspace = true }
    serde         = { workspace = true, optional = true }
    thiserror     = { workspace = true }
    tokio         = { workspace = true, features = ["sync"] }

[dev-dependencies]
    cargo-toolchain = { workspace = true }
    crossbeam       = { workspace = true }
    num_cpus        = { workspace = true }
    parking_lot     = { workspace = true }
    rustc_version   = { workspace = true }
    serde_json      = { workspace = true }
    tokio           = { workspace = true, default-features = false, features = ["macros", "rt", "rt-multi-thread", "time"] }
    trybuild        = { workspace = true }

[features]
    diagnostics = ["fieldx_derive/diagnostics"]
    send_guard  = ["parking_lot/send_guard"]
    serde       = ["fieldx_derive/serde", "parking_lot/serde", "dep:serde"]
