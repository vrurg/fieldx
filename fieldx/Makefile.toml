[tasks.test-feature-sets]
    category = "Test"
    clear = true
    description = "Test different feature sets"
    script = '''
!include_files ../makefile.ds

echo Testing feature sets with toolchain '${__FIELDX_DEFAULT_TOOLCHAIN__}'

features = array sync serde
feature_sets = generate_power_set ${features}

for fs in ${feature_sets}
    fs = array_join ${fs} ","
    echo "Testing feature set '${fs}'"
    set_env __FIELDX_FEATURE_SET__ ${fs}
    cm_run_task test-with-features
    cm_run_task test-with-features-async
end
'''
    script_runner = "@duckscript"
    toolchain = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.test-with-features-async]
    run_task = { name = ["test-async-lock", "test-async-tokio"], fork = true }

[tasks.test-async-lock]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-lock" }
    run_task = "test-with-features"

[tasks.test-async-tokio]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-tokio" }
    run_task = "test-with-features"

[tasks.test-with-features]
    category    = "Test"
    description = "Test code compilability with '${__FIELDX_FEATURE_SET__}' features"
    script_runner = "@duckscript"
    script = '''
exec --fail-on-error cargo +${__FIELDX_DEFAULT_TOOLCHAIN__} test --features ${__FIELDX_FEATURE_SET__}
exec --fail-on-error cargo +${__FIELDX_DEFAULT_TOOLCHAIN__} test --features ${__FIELDX_FEATURE_SET__} --examples
'''

# Compilation is useful for updating the error outputs per version+features
[tasks.test-compilation]
    category = "Test"
    clear = true
    description = "Test code compilability"
    script = '''
!include_files ../makefile.ds

echo Testing compilability with toolchain '${__FIELDX_DEFAULT_TOOLCHAIN__}'
cm_run_task compile-default-features

features = array sync serde
feature_sets = generate_power_set ${features}

for fs in ${feature_sets}
    fs = array_join ${fs} ","
    echo "Testing feature set '${fs}' on ${__FIELDX_DEFAULT_TOOLCHAIN__}"
    set_env __FIELDX_FEATURE_SET__ ${fs}
    cm_run_task compile-with-features
    cm_run_task compile-with-features-async
end
'''
    script_runner = "@duckscript"
    toolchain = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-default-features]
    condition   = { env_not = { __FIELDX_DEFAULT_TOOLCHAIN__ = "nightly" } }
    args        = ["test", "--test", "compilation", "--", "--nocapture"]
    category    = "Test"
    command     = "cargo"
    description = "Test code compilability with default feature set"
    toolchain   = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-with-features]
    condition   = { env_not = { __FIELDX_DEFAULT_TOOLCHAIN__ = "nightly" } }
    args        = ["test", "--test", "compilation", "--features", "${__FIELDX_FEATURE_SET__}", "--", "--nocapture"]
    category    = "Test"
    command     = "cargo"
    description = "Test code compilability with '${__FIELDX_FEATURE_SET__}' features"
    toolchain   = "${__FIELDX_DEFAULT_TOOLCHAIN__}"

[tasks.compile-with-features-async]
    run_task = { name = ["compile-async-lock", "compile-async-tokio"], fork = true }

[tasks.compile-async-lock]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-lock" }
    run_task = "compile-with-features"

[tasks.compile-async-tokio]
    env      = { __FIELDX_FEATURE_SET__ = "${__FIELDX_FEATURE_SET__},async-tokio" }
    run_task = "compile-with-features"
