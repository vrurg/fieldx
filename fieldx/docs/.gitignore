.markdownlint.json
doctest_cache
target-doc/
book/