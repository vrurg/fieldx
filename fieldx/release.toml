pre-release-hook = ['/bin/sh', '-c', '''
if [ "$CARGO_MAKE_TASK" != "publish" -a "$CARGO_MAKE_TASK" != "beta" ]; then
    echo "***ERROR*** Must be released with help of 'cargo make publish' in project workspace dir." 1>&2
    exit 1
else
    pushd $WORKSPACE_ROOT >/dev/null
    export CARGO_RELEASE_VERSION={{version}}
    if ! makers changelog; then
        echo "***ERROR*** Changelog generation failed." 1>&2
        exit 1
    fi
    popd
fi
''']

pre-release-replacements = [
    { file = "src/lib.rs", search = "(doc\\s*\\(html_root_url\\s*=\\s*\"https://docs.rs/.+?)/\\d+\\.\\d+\\.\\d.*/\"", replace = "$1/{{version}}/\"", prerelease = true },
]

enable-all-features = false
enable-features     = ["sync", "async-tokio", "serde", "clonable-lock"]
