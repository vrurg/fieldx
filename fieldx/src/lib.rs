pub mod errors;
pub mod traits;

pub use fieldx_derive::fxstruct;
pub use parking_lot::{MappedRwLockReadGuard, RwLock, RwLockReadGuard, RwLockUpgradableReadGuard, RwLockWriteGuard};
use std::{any, cell::RefCell, marker::PhantomData, sync::atomic::AtomicBool};
pub use std::{cell::OnceCell, fmt, sync::atomic::Ordering};
use traits::FXStructSync;

pub struct FXProxy<O, T>
where
    O: FXStructSync,
{
    value: RwLock<Option<T>>,
    is_set: AtomicBool,
    builder: RwLock<Option<fn(&O) -> T>>,
}

#[allow(private_bounds)]
pub struct FXWrLock<'a, O, T, FX>
where
    O: FXStructSync,
    FX: FXProxyCore<O, T>,
{
    lock: RefCell<RwLockWriteGuard<'a, Option<T>>>,
    fxproxy: &'a FX,
    _phantom: PhantomData<O>,
}

impl<O, T> Default for FXProxy<O, T>
where
    O: FXStructSync,
{
    fn default() -> Self {
        Self {
            value: RwLock::new(None),
            is_set: AtomicBool::new(false),
            builder: RwLock::new(None),
        }
    }
}

impl<O, T: fmt::Debug> fmt::Debug for FXProxy<O, T>
where
    O: FXStructSync,
{
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let vlock = self.value.read();
        formatter
            .debug_struct(any::type_name::<Self>())
            .field("value", &*vlock)
            .finish()
    }
}

impl<O, T> From<T> for FXProxy<O, T>
where
    O: FXStructSync,
    Self: FXProxyCore<O, T>,
{
    fn from(value: T) -> Self {
        Self {
            value: RwLock::new(Some(value)),
            is_set: AtomicBool::new(true),
            builder: RwLock::new(None),
        }
    }
}

pub trait FXProxyCore<O, T>
where
    O: FXStructSync,
    Self: Sized,
{
    fn builder(&self) -> &RwLock<Option<fn(&O) -> T>>;
    fn value(&self) -> &RwLock<Option<T>>;
    fn is_set_raw(&self) -> &AtomicBool;

    fn is_set(&self) -> bool {
        self.is_set_raw().load(Ordering::SeqCst)
    }

    fn proxy_setup(&self, builder: fn(&O) -> T) {
        *self.builder().write() = Some(builder);
        if !self.is_set() && (*self.value().read()).is_some() {
            self.is_set_raw().store(true, Ordering::SeqCst);
        }
    }

    fn read<'a>(&'a self, owner: &O) -> MappedRwLockReadGuard<'a, T> {
        let guard = self.value().upgradable_read();
        if (*guard).is_none() {
            let mut wguard = RwLockUpgradableReadGuard::upgrade(guard);
            // Still uninitialized? Means no other thread took care of it yet.
            if wguard.is_none() {
                // No value has been set yet
                match *self.builder().read() {
                    Some(ref builder_cb) => {
                        *wguard = Some((*builder_cb)(owner));
                        self.is_set_raw().store(true, Ordering::SeqCst);
                    }
                    None => panic!("Builder is not set"),
                }
            }
            return RwLockReadGuard::map(RwLockWriteGuard::downgrade(wguard), |data: &Option<T>| {
                data.as_ref().unwrap()
            });
        }
        RwLockReadGuard::map(RwLockUpgradableReadGuard::downgrade(guard), |data: &Option<T>| {
            data.as_ref().unwrap()
        })
    }

    fn write<'a>(&'a self) -> FXWrLock<'a, O, T, Self> {
        FXWrLock::<'a, O, T, Self>::new(self.value().write(), self)
    }

    fn clear_with_lock(&self, wguard: &mut RwLockWriteGuard<Option<T>>) -> Option<T> {
        self.is_set_raw().store(false, Ordering::SeqCst);
        wguard.take()
    }

    fn clear(&self) -> Option<T> {
        let mut wguard = self.value().write();
        self.clear_with_lock(&mut wguard)
    }
}

impl<O, T> FXProxyCore<O, T> for FXProxy<O, T>
where
    O: FXStructSync,
{
    #[inline]
    fn builder(&self) -> &RwLock<Option<fn(&O) -> T>> {
        &self.builder
    }

    #[inline]
    fn value(&self) -> &RwLock<Option<T>> {
        &self.value
    }

    #[inline]
    fn is_set_raw(&self) -> &AtomicBool {
        &self.is_set
    }
}

impl<O, T> FXProxy<O, T> where O: FXStructSync {}

#[allow(private_bounds)]
impl<'a, O, T, FX> FXWrLock<'a, O, T, FX>
where
    O: FXStructSync,
    FX: FXProxyCore<O, T>,
{
    pub fn new(lock: RwLockWriteGuard<'a, Option<T>>, fxproxy: &'a FX) -> Self
    where
        FX: FXProxyCore<O, T>,
    {
        let lock = RefCell::new(lock);
        Self {
            lock,
            fxproxy,
            _phantom: PhantomData,
        }
    }

    pub fn store(&mut self, value: T) -> Option<T> {
        self.fxproxy.is_set_raw().store(true, Ordering::Release);
        self.lock.borrow_mut().replace(value)
    }

    pub fn clear(&self) -> Option<T> {
        self.fxproxy.clear_with_lock(&mut *self.lock.borrow_mut())
    }
}
