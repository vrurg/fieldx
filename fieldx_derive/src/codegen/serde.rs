use super::{FXCGenContextual, FXFieldCtx, FXValueRepr};
use crate::helper::with_origin::FXOrig;
use darling::ast::NestedMeta;
use proc_macro2::{Span, TokenStream};
use quote::{quote, quote_spanned, ToTokens};
use syn::{spanned::Spanned, Meta};

pub(crate) trait FXCGenSerde<'f>: FXCGenContextual<'f> {
    // Field is an Option in the shadow struct if it is optional or lazy and has no default value
    fn is_serde_optional(&self, fctx: &FXFieldCtx) -> bool {
        fctx.is_optional() || fctx.is_lazy()
    }

    fn filter_shadow_attributes<'a>(&'a self, fctx: &'a FXFieldCtx) -> impl Iterator<Item = &'a syn::Attribute> {
        // Only use `serde` attribute and those listed in forward_attrs
        let serde_helper = fctx.serde().as_ref().or_else(|| self.ctx().args().serde().as_ref());
        fctx.attrs()
            .iter()
            .filter(move |a| a.path().is_ident("serde") || serde_helper.map_or(false, |sh| sh.accepts_attr(a)))
    }

    fn serde_derive_traits(&self) -> Vec<TokenStream> {
        let mut traits: Vec<TokenStream> = vec![];
        let ctx = self.ctx();
        if ctx.args().is_serde() {
            let serde_arg = ctx.args().serde().as_ref();
            let serde_helper = serde_arg.unwrap();
            let serde_helper_span = serde_helper.to_token_stream().span();

            if serde_helper.needs_serialize().unwrap_or(true) {
                traits.push(quote_spanned![serde_helper_span=> Serialize]);
            }
            if serde_helper.needs_deserialize().unwrap_or(true) {
                traits.push(quote_spanned![serde_helper_span=> Deserialize]);
            }
        }
        return traits;
    }

    fn serde_skip_toks(&self, field_ctx: &FXFieldCtx) -> TokenStream {
        // Don't skip a field if:
        // - no `serde` argument
        // - it is not `serde(off)`
        // - and no more than one of `deserialize` or `serialize` is `off`
        if self.ctx().args().is_serde() {
            let helper_span = field_ctx
                .serde()
                .as_ref()
                .and_then(|sh| sh.orig())
                .map_or(Span::call_site(), |s| s.span());
            if !field_ctx.is_serde() {
                return quote_spanned!(helper_span=> skip );
            }
            if !field_ctx.needs_serialize() {
                return quote_spanned!(helper_span=> skip_serializing );
            }
            if !field_ctx.needs_deserialize() {
                return quote_spanned!(helper_span=> skip_deserializing );
            }
        }
        quote![]
    }

    fn serde_struct_attribute(&self) -> darling::Result<()> {
        let ctx = self.ctx();
        let args = ctx.args();

        ctx.add_attr_from(if args.is_serde() {
            let mut serde_args: Vec<TokenStream> = vec![];

            let serde_helper = args.serde().as_ref().unwrap();
            let generics = ctx.input().generics();
            let shadow_ident = format!("{}{}", ctx.shadow_ident(), generics.to_token_stream());

            if serde_helper.needs_deserialize().unwrap_or(true) {
                serde_args.push(quote![from = #shadow_ident]);
            }
            if serde_helper.needs_serialize().unwrap_or(true) {
                serde_args.push(quote![into = #shadow_ident]);
            }

            if serde_args.len() > 0 {
                quote![ #[serde( #( #serde_args ),*  )] ]
            }
            else {
                quote![]
            }
        }
        else {
            quote![]
        });

        Ok(())
    }

    fn serde_field_attribute(&self, fctx: &FXFieldCtx) -> darling::Result<TokenStream> {
        let ctx = self.ctx();
        Ok(if ctx.args().is_serde() {
            let skip_toks = self.serde_skip_toks(fctx);
            let mut serde_attr_args = vec![];

            if !skip_toks.is_empty() {
                serde_attr_args.push(skip_toks);
            }

            let mut default_arg = None;

            if let Some(serde_helper) = fctx.serde().as_ref() {
                if serde_helper.has_default() {
                    // Safe because of has_default()
                    let default_value = serde_helper.default_value_raw().unwrap();
                    let span = default_value.orig().span();

                    if default_value.has_value() {
                        let serde_default_str: String = if default_value.is_str() {
                            default_value.try_into()?
                        }
                        else {
                            let struct_ident = ctx.input_ident();
                            let (_, generics, _) = ctx.input().generics().split_for_impl();
                            let default_fn_ident = self.serde_field_default_fn(fctx)?;

                            format!(
                                "{}{}::{}",
                                struct_ident,
                                generics.as_turbofish().to_token_stream(),
                                default_fn_ident
                            )
                        };

                        default_arg = Some(quote_spanned![span=> default = #serde_default_str]);
                    }
                    else {
                        default_arg = Some(quote_spanned![span=> default]);
                    }
                }
            }

            if let Some(default_arg) = default_arg {
                serde_attr_args.push(default_arg);
            }

            if let Some(base_name) = fctx.base_name() {
                let span = base_name.span();
                let base_name = base_name.to_string();
                serde_attr_args.push(quote_spanned![span=> rename = #base_name]);
            }

            if serde_attr_args.is_empty() {
                quote![]
            }
            else {
                quote! [ #[serde( #( #serde_attr_args ),* )] ]
            }
        }
        else {
            quote![]
        })
    }

    fn serde_shadow_field_type(&self, fctx: &FXFieldCtx) -> TokenStream {
        let ty = fctx.ty_tok().clone();
        if self.is_serde_optional(fctx) {
            quote![ ::std::option::Option<#ty> ]
        }
        else {
            quote![ #ty ]
        }
    }

    fn serde_shadow_field_value(&self, fctx: &FXFieldCtx, value: TokenStream) -> TokenStream {
        if self.is_serde_optional(fctx) {
            quote![ ::std::option::Option::Some(#value) ]
        }
        else {
            value
        }
    }

    fn serde_shadow_field(&self, fctx: &FXFieldCtx) {
        let ident = fctx.ident_tok();
        let attrs = self.filter_shadow_attributes(fctx);
        let serde_attr = self.ok_or_empty(self.serde_field_attribute(fctx));
        let user_attrs = fctx
            .serde()
            .as_ref()
            .and_then(|serde_helper| serde_helper.attributes().as_ref())
            .or_else(|| self.ctx().args().attributes().as_ref());
        let ty = self.serde_shadow_field_type(fctx);

        self.add_shadow_field_decl(quote_spanned! [*fctx.span()=>
            #serde_attr
            #user_attrs
            #( #attrs )*
            #ident: #ty
        ]);
    }

    fn serde_shadow_field_default(&self, fctx: &FXFieldCtx) {
        let mut default_tok = self.fixup_self_type(
            self.ok_or_else(self.field_default_value(fctx), || FXValueRepr::None)
                .unwrap_or(quote![::std::default::Default::default()]),
        );

        if self.is_serde_optional(fctx) {
            default_tok = self.serde_shadow_field_value(fctx, default_tok);
        }

        let field_ident = fctx.ident_tok();

        self.add_shadow_default_decl(quote![ #field_ident: #default_tok ]);
    }

    fn serde_shadow_struct(&self) -> darling::Result<()> {
        let ctx = self.ctx();
        let args = ctx.args();
        if args.is_serde() {
            let serde_helper = args.serde().as_ref().unwrap();
            let shadow_ident = ctx.shadow_ident();
            let fields = self.shadow_fields();
            let mut attrs = vec![];
            let derive_attr = self.derive_toks(&self.serde_derive_traits());
            let shadow_defaults = self.shadow_defaults();
            let generics = ctx.input().generics();
            let where_clause = &generics.where_clause;
            let vis = serde_helper.public_mode().map(|pm| pm.to_token_stream());
            let user_attributes = serde_helper.attributes();

            attrs.push(derive_attr);

            if let Some(default_attr_arg) = self.serde_shadow_default_fn()? {
                attrs.push(quote![#[serde(#default_attr_arg)]]);
            }

            ctx.tokens_extend(quote![
                #( #attrs )*
                #user_attributes
                #vis struct #shadow_ident #generics #where_clause {
                    #( #fields ),*
                }

                // #default_impl

                impl #generics ::std::default::Default for #shadow_ident #generics #where_clause {
                    fn default() -> Self {
                        Self {
                            #( #shadow_defaults ),*
                        }
                    }
                }
            ]);
        }

        Ok(())
    }

    // Impl From for the shadow struct
    fn serde_struct_from_shadow(&'f self) {
        let ctx = self.ctx();
        let args = ctx.args();
        if args.is_serde() && args.needs_deserialize() {
            let shadow_ident = ctx.shadow_ident();
            let struct_ident = ctx.input_ident();
            let shadow_var = ctx.shadow_var_ident();
            let mut fields = vec![];
            let generics = ctx.input().generics();
            let where_clause = &generics.where_clause;

            for field in ctx.input().fields() {
                let fctx = self.field_ctx(&field);
                if let Ok(fctx) = fctx {
                    if fctx.is_serde() && fctx.needs_deserialize() {
                        let field_ident = fctx.ident_tok();
                        match self.field_from_shadow(&fctx) {
                            Ok(fetch_shadow_field) => fields.push(quote_spanned![*fctx.span()=>
                                #field_ident: #fetch_shadow_field
                            ]),
                            Err(err) => ctx.push_error(err),
                        }
                    }
                }
                else {
                    ctx.push_error(fctx.unwrap_err())
                }
            }

            ctx.tokens_extend(quote![
                impl #generics ::std::convert::From<#shadow_ident #generics> for #struct_ident #generics #where_clause {
                    fn from(#shadow_var: #shadow_ident #generics) -> Self {
                        Self {
                            #( #fields, )*
                            .. Self::default()
                        }
                    }
                }
            ]);
        }
    }

    fn serde_struct_into_shadow(&'f self) {
        let ctx = self.ctx();
        let args = ctx.args();
        if args.is_serde() && args.needs_serialize() {
            let shadow_ident = ctx.shadow_ident();
            let struct_ident = ctx.input_ident();
            let mut fields = vec![];
            let mut lazy_inits = vec![];
            let me_var = ctx.me_var_ident();
            let generics = ctx.input().generics();
            let where_clause = &generics.where_clause;

            for field in ctx.input().fields() {
                let fctx = self.field_ctx(&field);
                if let Ok(fctx) = fctx {
                    if fctx.is_serde() && fctx.needs_serialize() {
                        let field_ident = fctx.ident_tok();
                        let is_lazy = fctx.is_lazy();
                        let fetch_struct_field = match self.field_from_struct(&fctx) {
                            Ok(f) => f,
                            Err(err) => {
                                ctx.push_error(err);
                                continue;
                            }
                        };

                        if is_lazy {
                            let init_call =
                                self.ok_or_empty(self.field_lazy_initializer(&fctx, Some(quote![(&#me_var)])));
                            lazy_inits.push(quote![ let _ = #init_call; ]);
                        }

                        fields.push(quote_spanned![*fctx.span()=> #field_ident: #fetch_struct_field ]);
                    }
                }
                else {
                    ctx.push_error(fctx.unwrap_err())
                }
            }

            ctx.tokens_extend(quote![
                impl #generics ::std::convert::From<#struct_ident #generics> for #shadow_ident #generics #where_clause {
                    fn from(mut #me_var: #struct_ident #generics) -> Self {
                        #( #lazy_inits )*
                        Self {
                            #( #fields, )*
                        }
                    }
                }
            ])
        }
    }

    fn serde_prepare_struct(&'f self) {
        for field in self.input().fields() {
            let Ok(fctx) = self.field_ctx(field)
            else {
                continue;
            };

            if fctx.is_serde() {
                self.serde_shadow_field(&fctx);
                self.serde_shadow_field_default(&fctx);
            }
        }

        self.ctx().add_attr_from(self.derive_toks(&self.serde_derive_traits()));

        self.ok_or_record(self.serde_struct_attribute());
    }

    fn serde_rewrite_struct(&'f self) {
        self.ok_or_record(self.serde_shadow_struct());
        self.serde_struct_from_shadow();
        self.serde_struct_into_shadow();
    }

    fn serde_shadow_default_fn(&self) -> darling::Result<Option<TokenStream>> {
        let ctx = self.ctx();

        let Some(serde_helper) = ctx.args().serde()
        else {
            return Ok(None);
        };

        if serde_helper.has_default() {
            let default_value = serde_helper.default_value_raw().unwrap();
            let span = default_value.orig().span();

            if default_value.has_value() {
                let serde_default: TokenStream = if default_value.is_str() {
                    let default_str: String = default_value.try_into()?;
                    let expr: syn::ExprPath = syn::parse_str(&default_str).map_err(|err| {
                        darling::Error::custom(format!("Invalid default string: {}", err)).with_span(&span)
                    })?;
                    quote![#expr()]
                }
                else {
                    let default_code = default_value.value().as_ref().unwrap();
                    if let NestedMeta::Meta(Meta::NameValue(_)) = default_code {
                        let err = darling::Error::custom(format!("Unexpected kind of argument")).with_span(&span);
                        #[cfg(feature = "diagnostics")]
                        let err = err.note(format!(
                            "{}\n{}\n{}",
                            "Consider using a string, as with serde `default`: \"Type::function\"`",
                            "                                       or a path: `Type::static_or_constant`",
                            "                       or a call-like expression: `Type::function()`"
                        ));
                        return Err(err);
                    }
                    quote![#default_code]
                };

                let generics = ctx.input().generics();
                let shadow_ident = ctx.shadow_ident();
                let fn_ident = ctx.unique_ident_pfx(&format!("{}_default", shadow_ident.to_string()));
                self.add_method_decl(quote![
                    fn #fn_ident() -> #shadow_ident #generics {
                        #serde_default.into()
                    }
                ]);

                let default_str = format!("{}::{}", ctx.input_ident(), fn_ident);

                return Ok(Some(quote_spanned![span=> default = #default_str]));
            }
            else {
                return Ok(Some(quote_spanned![span=> default]));
            }
        }

        Ok(None)
    }

    fn serde_field_default_fn(&self, fctx: &FXFieldCtx) -> darling::Result<syn::Ident> {
        let fn_ident = fctx.default_fn_ident()?;
        let field_type = self.serde_shadow_field_type(fctx);
        let Some(serde_helper) = fctx.serde()
        else {
            return Err(darling::Error::custom(format!(
                "Can't generate default function for non-serde field {}",
                fctx.ident_str()
            )));
        };
        let Some(serde_default) = serde_helper
            .default_value()
            .map(|dv| self.serde_shadow_field_value(fctx, dv.to_token_stream()))
        else {
            return Err(darling::Error::custom(format!(
                "There is no serde 'default' for field {}",
                fctx.ident_str()
            )));
        };

        self.add_method_decl(quote![
            fn #fn_ident() -> #field_type {
                #serde_default
            }
        ]);
        Ok(fn_ident.clone())
    }
}

impl<'f, T> FXCGenSerde<'f> for T where T: FXCGenContextual<'f> {}
