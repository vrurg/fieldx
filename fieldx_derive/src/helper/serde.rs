use super::{FXHelper, FXHelperTrait, FXNestingAttr, FromNestAttr};
use crate::util::set_literals;
use darling::{
    util::{Flag, PathList},
    FromMeta,
};
use getset::Getters;
use syn::Lit;

#[derive(FromMeta, Clone, Default, Debug, Getters)]
#[darling(default)]
pub(crate) struct FXSerdeDefault {
    off:   Flag,
    #[getset(get="pub")]
    value: Option<String>,
}

#[derive(Default, Debug, Getters, FromMeta, Clone)]
#[getset(get = "pub(crate)")]
pub(crate) struct FXSerdeHelper<const SHADOW_NAME: bool = false> {
    off:           Flag,
    serialize:     Option<FXHelper<true>>,
    deserialize:   Option<FXHelper<true>>,
    // Attributes of the original struct to be used with the shadow struct.
    forward_attrs: Option<PathList>,
    #[darling(rename = "default")]
    default_value: Option<FXNestingAttr<FXSerdeDefault>>,
    // Name of the new type to be used for deserialization. Normally its <ident>Shadow
    shadow_name:   Option<String>,
}

impl FromNestAttr for FXSerdeHelper {
    set_literals! {serde}

    fn for_keyword() -> darling::Result<Self> {
        Ok(Self::default())
    }
}

impl FXSerdeHelper {
    pub(crate) fn is_true(&self) -> bool {
        !self.off.is_present()
    }

    pub(crate) fn needs_serialize(&self) -> Option<bool> {
        self.serialize
            .as_ref()
            .map(|s| s.is_true())
            .or_else(|| self.deserialize.as_ref().map(|d| !d.is_true()))
    }

    pub(crate) fn needs_deserialize(&self) -> Option<bool> {
        self.deserialize
            .as_ref()
            .map(|d| d.is_true())
            .or_else(|| self.serialize.as_ref().map(|s| !s.is_true()))
    }

    pub(crate) fn is_serde(&self) -> Option<bool> {
        // Consider as Some(true) if not `serde(off)` or any of `serialize` or `deserialize` is defined and not both are
        // `off`. I.e. since `serde(deserialize(off))` implies `serialize` being `on` then the outcome is `Some(true)`.
        if self.is_true() {
            let is_serialize = self.serialize.as_ref().map(|s| s.is_true());
            let is_deserialize = self.deserialize.as_ref().map(|d| d.is_true());

            if is_serialize.is_none() && is_deserialize.is_none() {
                return None;
            }

            Some(is_serialize.unwrap_or(true) || is_deserialize.unwrap_or(true))
        }
        else {
            Some(false)
        }
    }

    pub(crate) fn accepts_attr(&self, attr: &syn::Attribute) -> bool {
        self.forward_attrs.as_ref().map_or(true, |fa| fa.contains(attr.path()))
    }

    pub(crate) fn has_default(&self) -> bool {
        self.default_value.as_ref().map_or(false, |d| d.is_true())
    }

    // pub(crate) fn default_value(&self) -> Option<&String> {
    //     self.default_value.as_ref().and_then(|d| d.value().as_ref())
    // }
}

impl FromNestAttr for FXSerdeDefault {
    set_literals! {default, ..1 => value as Lit::Str}

    fn for_keyword() -> darling::Result<Self> {
        Ok(Self::default())
    }
}

impl FXSerdeDefault {
    fn is_true(&self) -> bool {
        !self.off.is_present()
    }
}