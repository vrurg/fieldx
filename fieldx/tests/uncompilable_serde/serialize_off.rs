#![cfg(feature = "serde")]
// Make sure shadow struct is not produced if both serialize and deserialize arguments of struct `serde` argument are
// disabled.
use fieldx::fxstruct;
use serde::{Deserialize, Serialize};
use serde_json;

#[fxstruct(sync, serde(serialize(off)))]
struct Foo {
    v: &'static str,
}

fn main() {
    let f = Foo { v: "whatever" };
    let _json = serde_json::to_string(&foo);
}
