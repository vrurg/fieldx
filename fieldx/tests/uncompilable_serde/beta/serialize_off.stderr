error[E0425]: cannot find value `foo` in this scope
  --> tests/uncompilable_serde/serialize_off.rs:15:40
   |
15 |     let _json = serde_json::to_string(&foo);
   |                                        ^^^ not found in this scope

warning: unused import: `Serialize`
 --> tests/uncompilable_serde/serialize_off.rs:5:26
  |
5 | use serde::{Deserialize, Serialize};
  |                          ^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
